contrib/transfer_hierarchy.lua

```diff
@@ -0,0 +1,347 @@
+   end
+   if commonSegments == nil then
+      return prefix
+   end
+   if commonSegments[#commonSegments] == PATH_SEPARATOR then
+      table.remove(commonSegments, #commonSegments)
+   end
+   rv = prefix .. table.concat(commonSegments)
+   return rv
+end
+
+local function calculateRoot()
+   films = {}
+   for _,img in ipairs(darktable.gui.action_images) do
+      films[img.film] = true
+   end
+   return findRootPath(films), films
+end
+
+local function doCalculate()
+   local rootPath = calculateRoot()
+   if rootPath ~= nil then
+      sourceTextBox.text = rootPath
+      local sourceBase = darktable.preferences.read(LIB_ID, "source_base", "directory")
+      local destBase = darktable.preferences.read(LIB_ID, "destination_base", "directory")
+      if sourceBase ~= nil and sourceBase ~= "" and
+	 destBase ~= nil and destBase ~= "" and
+	 string.sub(rootPath, 1, #sourceBase) == sourceBase then
+      	 destinationTextBox.text = destBase .. string.sub(rootPath, #sourceBase+1)
+      end
+   end
+end
+
+local function stopTransfer(transferJob)
+   transferJob.valid = false
+end
+
+local function doTransfer(transferFunc)
+    rootPath, films = calculateRoot()
+    if rootPath ~= sourceTextBox.text then
+       darktable.print("transfer hierarchy: ERROR: existing root is out of sync -- click 'calculate' to update")
+       return
+    end
+    if destinationTextBox.text == "" then
+       darktable.print("transfer hierarchy: ERROR: destination not specified")
+       return
+    end
+    local sourceBase = sourceTextBox.text
+    local destBase = destinationTextBox.text
+    local destFilms = {}
+    for film, _ in pairs(films) do
+       films[film] = destBase .. string.sub(film.path, #sourceBase+1)
+       if not pathExists(films[film]) then
+           if createDirectory(films[film]) == nil then
+	       darktable.print("transfer hierarchy: ERROR: could not create directory: " .. films[film])
+	       return
+	   end
+       end
+       if not pathIsDirectory(films[film]) then
+           darktable.print("transfer hierarchy: ERROR: not a directory: " .. films[film])
+	   return
+       end
+       destFilms[film] = darktable.films.new(films[film])
+       if destFilms[film] == nil then
+           darktable.print("transfer hierarchy: ERROR: could not create film: " .. film.path)
+       end
+    end
+
+    local job = darktable.gui.create_job(string.format("transfer hierarchy (%d image" .. (#(darktable.gui.action_images) == 1 and "" or "s") .. ")", #(darktable.gui.action_images)), true, stopTransfer)
+    job.percent = 0.0
+    local pctIncrement = 1.0 / #(darktable.gui.action_images)
+    for _,img in ipairs(darktable.gui.action_images) do
+       if job.valid then
+    	  destFilm = destFilms[img.film]
+	  transferFunc(img, destFilm)
+	  job.percent = job.percent + pctIncrement
+       end
+    end
+    job.valid = false
+    local filterRules = darktable.gui.libs.collect.filter()
+    darktable.gui.libs.collect.filter(filterRules)
```

Comment:
I did a copy from a collection to a new collection.  After the copy I expected to end up at the destination, but the existing root just got refreshed and set to the top of the collection.  Would it be better to end up at the destination after a move or copy?

Reply:
Keeping the view on the source was done on account of there being no single "destination" directory for the images, so the only possible collection change would be to a collect-by-film-roll on the destination root, which in my case is an already-well-populated archive. Finding the freshly moved or copied images among all the images there is like finding a needle in a haystack.

Reply:
That makes sense